extensions/charts/cisbenchmark.go

```diff
@@ -0,0 +1,298 @@
+			} else if chart == nil {
+				return true, nil
+			}
+
+			return false, nil
+		})
+		if err != nil {
+			return err
+		}
+
+		steveclient, err := client.Steve.ProxyDownstream(installOptions.Cluster.ID)
+		if err != nil {
+			return err
+		}
+
+		namespaceClient := steveclient.SteveType(namespaces.NamespaceSteveType)
+
+		namespace, err := namespaceClient.ByID(CISBenchmarkNamespace)
+		if err != nil {
+			return err
+		}
+
+		err = namespaceClient.Delete(namespace)
+		if err != nil {
+			return err
+		}
+
+		adminClient, err := rancher.NewClient(client.RancherConfig.AdminToken, client.Session)
+		if err != nil {
+			return err
+		}
+
+		adminDynamicClient, err := adminClient.GetDownStreamClusterClient(installOptions.Cluster.ID)
+		if err != nil {
+			return err
+		}
+
+		adminNamespaceResource := adminDynamicClient.Resource(kubenamespaces.NamespaceGroupVersionResource).Namespace("")
+
+		watchNamespaceInterface, err := adminNamespaceResource.Watch(context.TODO(), metav1.ListOptions{
+			FieldSelector:  "metadata.name=" + CISBenchmarkNamespace,
+			TimeoutSeconds: &defaults.WatchTimeoutSeconds,
+		})
+		if err != nil {
+			return err
+		}
+
+		return wait.WatchWait(watchNamespaceInterface, func(event watch.Event) (ready bool, err error) {
+			if event.Type == watch.Deleted {
+				return true, nil
+			}
+
+			return false, nil
+		})
+	})
+
+	err = catalogClient.InstallChart(chartInstallAction, catalog.RancherChartRepo)
+	if err != nil {
+		return err
+	}
+
+	watchAppInterface, err := catalogClient.Apps(CISBenchmarkNamespace).Watch(context.TODO(), metav1.ListOptions{
+		FieldSelector:  "metadata.name=" + CISBenchmarkName,
+		TimeoutSeconds: &defaults.WatchTimeoutSeconds,
+	})
+	if err != nil {
+		return err
+	}
+
+	err = wait.WatchWait(watchAppInterface, func(event watch.Event) (ready bool, err error) {
+		app := event.Object.(*catalogv1.App)
+
+		state := app.Status.Summary.State
+		if state == string(catalogv1.StatusDeployed) {
+			return true, nil
+		}
+
+		return false, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// newCISBenchmarkChartInstallAction is a private helper function that returns chart install action with CIS benchmark and payload options.
+func newCISBenchmarkChartInstallAction(p *payloadOpts) (*types.ChartInstallAction, error) {
+	chartInstall := newChartInstall(p.Name, p.Version, p.Cluster.ID, p.Cluster.Name, p.Host, rancherChartsName, p.ProjectID, p.DefaultRegistry, nil)
+	chartInstallCRD := newChartInstall(p.Name+"-crd", p.Version, p.Cluster.ID, p.Cluster.Name, p.Host, rancherChartsName, p.ProjectID, p.DefaultRegistry, nil)
+	chartInstalls := []types.ChartInstall{*chartInstallCRD, *chartInstall}
+
+	chartInstallAction := newChartInstallAction(p.Namespace, p.ProjectID, chartInstalls)
+
+	return chartInstallAction, nil
+}
+
+// UpgradeCISBenchmarkChart is a helper function that upgrades the cis-benchmark chart.
+func UpgradeCISBenchmarkChart(client *rancher.Client, installOptions *InstallOptions) error {
```

Comment:
do you run upgrade tests anywhere yet? or is this just for support later?

Reply:
This is support for later, correct.